### PR TITLE
Remove additional weight prefetching all gathers 

### PR DESCRIPTION
**Pull request by NuojCheng**
# Description

This PR does following 2 things:

- Enable real buffered sliding window (BSW) for circular pipeline parallelism: we only need to gather one layer's weight and scan carry the old weight without incurring memory burden, see description in my doc go/maxtext-spmd-p-2026.
- Add context paralleism (CP) support for custom mesh and logical rule as well as weight prefetching of circular pipeline parallelism.

# Tests

Please describe how you tested this change, and include any instructions and/or
commands to reproduce.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/layers/pipeline.py [hunk 1]
```diff
@@ -1006,8 +1007,12 @@ def gather_microbatch_inputs_vmap(self, xs, ids, ids_dim):
 
     def _gather_one(x, i):
       idx = tuple(i if d == ids_dim else slice(None) for d in range(x.ndim))
-      replicated_sharding = NamedSharding(self.mesh, P())
-      return x.at[idx].get(out_sharding=replicated_sharding)
+      positions_sharding = (
```
[the remark this answers is not in the document]
Comment (gobbleturk) [reply src/maxtext/layers/pipeline.py:1010]: I guess mr gemini review is wrong here? what are the axes of positions sharding?

File: src/maxtext/layers/pipeline.py [hunk 2]
```diff
@@ -1168,7 +1173,7 @@ def from_repeat_weights_to_bsw(
       self,
       repeat_weights,
       physical_partition_spec,
-      axes_to_gather=("fsdp", "fsdp_transpose", "expert"),  # three major FSDP-like axes
+      axes_to_gather=("fsdp", "fsdp_transpose", "context", "expert"),  # three major FSDP-like axes
```
[the remark this answers is not in the document]
Comment (gobbleturk) [reply src/maxtext/layers/pipeline.py]: AGI +1

File: src/maxtext/utils/pipeline_utils.py [hunk 3]
```diff
@@ -281,38 +274,68 @@ def create_rematerialized_pipeline_stage(
...
-    bsw = model.weight_prefetching(weights, physical_partition_spec, loop_state["loop_iteration"])
-
-    if model.config.scan_pipeline_iterations:
-      scan_microbatches_fn = create_gradient_accumulation_scan(
-          model=model,
-          length=length,
-          deterministic=deterministic,
-          model_mode=model_mode,
-          logical_partition_spec=logical_partition_spec,
+  def execute_pipeline_stage_outer(model, carry):
```
Comment (gobbleturk) [inline src/maxtext/utils/pipeline_utils.py]: can we have a more descriptive name? Is this execute_pipeline_stage_layer (all microbatches corresponding to one layer?)

File: src/maxtext/utils/pipeline_utils.py [hunk 4]
```diff
@@ -281,38 +274,68 @@ def create_rematerialized_pipeline_stage(
...
+          remat_weight_prefetching,
+          physical_partition_spec=physical_partition_spec,
+          loop_iteration=loop_state["loop_iteration"],
+      )
+      remat_weight_prefetching_t = jax.linear_transpose(
+          p_remat_weight_prefetching,
+          pipeline_weights,
+      )
+      (loop_state, bsw), scan_fn_vjp = jax.vjp(scan_microbatches_fn, loop_state, bsw, positions, segment_ids)
+      w_curr, w_next = bsw
```
[the remark this answers is not in the document]
Comment (gobbleturk) [reply src/maxtext/utils/pipeline_utils.py]: I agree with the vibes that this is complex code that is not easy to maintain (bus number = nuojin)

I'm not sure how to best increase readability here, perhaps you can include a link to jax's vjp doc? Is there a better name than scan_fn_vjp we can use - I guess scan_fn_vjp is a function that implements the backward pass?

[the remark this answers is not in the document]
Comment (gobbleturk) [reply src/maxtext/utils/pipeline_utils.py]: What is the first argument "residuals" in execute_pipeline_stage_custom_bwd

File: src/maxtext/layers/pipeline.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (NuojCheng) [reply src/maxtext/layers/pipeline.py:1010]: gemini does not understand explicit sharding too bad

File: src/maxtext/utils/pipeline_utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (NuojCheng) [reply src/maxtext/utils/pipeline_utils.py]: Added some comments to clarify! Residuals play as checkpoint that explicitly deliver from forward pass, and they will be saved in devices.

File: src/maxtext/utils/pipeline_utils.py (hunk 3 above)
Comment (NuojCheng) [reply src/maxtext/utils/pipeline_utils.py]: update it as `execute_pipeline_stage_flax` and the function inside as `execute_pipeline_stage_pure`

**Comment by NuojCheng** [comment]
> This is awesome!
> 
> Is there an existing test that protects the correctness of this? If not we definitely should add one in this PR

The loss and gradients correctness are protected by this test: https://github.com/AI-Hypercomputer/maxtext/blob/60dbacdf9d0fea44db44345d77d3919b9b9e9b4b/tests/unit/pipeline_parallelism_test.py#L283-L299
